pkg/ottl/ottlfuncs/func_parse_key_value.go

```diff
@@ -0,0 +1,88 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal/parseutils"
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type ParseKeyValueArguments[K any] struct {
+	Target        ottl.StringGetter[K]
+	Delimiter     ottl.Optional[string]
+	PairDelimiter ottl.Optional[string]
+}
+
+func NewParseKeyValueFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("ParseKeyValue", &ParseKeyValueArguments[K]{}, createParseKeyValueFunction[K])
+}
+
+func createParseKeyValueFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ParseKeyValueArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("ParseKeyValueFactory args must be of type *ParseKeyValueArguments[K]")
+	}
+
+	return parseKeyValue[K](args.Target, args.Delimiter, args.PairDelimiter)
+}
+
+func parseKeyValue[K any](target ottl.StringGetter[K], d ottl.Optional[string], p ottl.Optional[string]) (ottl.ExprFunc[K], error) {
+	delimiter := "="
+	if !d.IsEmpty() {
+		if d.Get() == "" {
+			return nil, fmt.Errorf("delimiter cannot be set to an empty string")
+		}
+		delimiter = d.Get()
+	}
+
+	pairDelimiter := " "
+	if !p.IsEmpty() {
+		if p.Get() == "" {
+			return nil, fmt.Errorf("pair delimiter cannot be set to an empty string")
+		}
+		pairDelimiter = p.Get()
+	}
+
+	if pairDelimiter == delimiter {
+		return nil, fmt.Errorf("pair delimiter %q cannot be equal to delimiter %q", pairDelimiter, delimiter)
+	}
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		source, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, err
+		}
+
+		if source == "" {
+			return nil, fmt.Errorf("cannot parse from empty target")
+		}
+
+		pairs, err := parseutils.SplitString(source, pairDelimiter)
+		if err != nil {
+			return nil, fmt.Errorf("splitting source %q into pairs failed: %w", source, err)
+		}
+
+		parsed := make(map[string]any)
+		for _, p := range pairs {
+			pair := strings.SplitN(p, delimiter, 2)
+			if len(pair) != 2 {
+				return nil, fmt.Errorf("cannot split %q into 2 items, got %d item(s)", p, len(pair))
+			}
+			key := strings.TrimSpace(pair[0])
+			value := strings.TrimSpace(pair[1])
+			parsed[key] = value
+		}
```

Comment:
I think this bit of code is the same as https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6f1b1d2a5c2a690e138feac4712c7af01b65f2a4/pkg/stanza/operator/parser/keyvalue/keyvalue.go#L106-L120, can it also be moved to `coreinternal`?

Reply:
Yea definitely, only question I have is how do we want to handle the error that occurs when the delimiter isn't found in the pair string? In Stanza it parses all pairs before returning where as the current OTTL implementation returns as soon as an error is encountered.

Reply:
I think following stanza here is ok.  We'll return a detailed error, which is helpful.

Reply:
I'm fine just matching Stanza's behavior. If in the future we want to change it in one place it probably makes sense to change it in both.